gpu-miner.cl

```diff
@@ -187,53 +85,29 @@ static int blake2b_compress( __private blake2b_state *S, __private const uchar b
 	ROUND( 9 );
 	ROUND( 10 );
 	ROUND( 11 );
-
+	uchar headerHash[64];
 	for( i = 0; i < 8; ++i )
-		S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
+		iv[i] = iv[i] ^ v[i] ^ v[i + 8];
+	for( int i = 0; i < 8; ++i ) // Output full hash to temp buffer
+		*(ulong*)(headerHash + 8 * i) = iv[i];
 
+	// BLAKE2B END
 #undef G
 #undef ROUND
-	return 0;
-}
-
-// inlen, at least, should be ulong. Others can be size_t.
-int blake2b( __private uchar *out, __private uchar *in )
-{
-	private blake2b_state S[1];
-
-	clmemset( S, 0, sizeof( blake2b_state ) );
-	for( int i = 0; i < 8; ++i ) S->h[i] = blake2b_IV[i];
-	S->h[0] ^= 0x0000000001010020UL;
-
-	ulong inlen = 80;
-	size_t left = S->buflen;
-	size_t fill = 2 * BLAKE2B_BLOCKBYTES - left;
 
-	if( inlen > fill )
-	{
-		clmemcpy( S->buf + left, in, fill ); // Fill buffer
-		S->buflen += fill;
-		blake2b_compress( S, S->buf ); // Compress
-		clmemcpy( S->buf, S->buf + BLAKE2B_BLOCKBYTES, BLAKE2B_BLOCKBYTES ); // Shift buffer left
-		S->buflen -= BLAKE2B_BLOCKBYTES;
+	// Compare hash to target
+	i = 0;
+	while (target[i] == headerHash[i]) {
+		i++;
```

Comment:
you can use `memcmp` here, as in https://github.com/NebulousLabs/Sia-GPU-Miner/blob/master/gpu-miner.c#L120

`memcmp` is safer than this loop anyway, since in the (extremely unlikely) event that `target == headerHash`, you'll get an out-of-bounds error.